src/node.c

```diff
@@ -145,6 +145,42 @@ cmark_node *cmark_node_new(cmark_node_type type) {
   return cmark_node_new_with_mem(type, &DEFAULT_MEM_ALLOCATOR);
 }
 
+int cmake_node_convert_type(cmark_node *node, cmark_node_type type) {
+  // Only need to validate children if we have any
+  if (node->first_child) {
+    enum node_class
+      old = S_compute_allowed_child(node->type),
+      new = S_compute_allowed_child(type);
+    switch (new) {
+    case NODE_CLASS_ANY:
+      // Doesn't matter what the old type is
+      break;
+
+    case NODE_CLASS_NONE:
+      // New type doesn't allow children, but have children
+      return 0;
+
+    case NODE_CLASS_BLOCK:
+    case NODE_CLASS_ITEM:
+    case NODE_CLASS_INLINE:
+      // Disjoint equiv classes, must match exactly
+      if (old != new)
+        return 0;
+      else
+        break;
+
+    // Handled all cases already
+    default:
+      assert(false);
+    }
+  }
+
+  // Success path
+  node->type = type;
+  S_cmark_node_init(node);
+  return 1;
```

Comment:
For some node types, you have to free old content before reinitializing the node, see `cmark_node_free`.

Reply:
Good catch, thanks!

Reply:
That is now fixed.